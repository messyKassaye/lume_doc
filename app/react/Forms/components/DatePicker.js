import { connect } from 'react-redux';
import 'react-datepicker/dist/react-datepicker.css';

import DatePickerComponent from 'react-datepicker';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import moment from 'moment';

const removeOffset = (useTimezone, value) => {
  let datePickerValue = null;
  if (value) {
    const newValue = moment.utc(value * 1000);

    if (!useTimezone) {
      newValue.subtract(moment().utcOffset(), 'minute');
    }

    datePickerValue = parseInt(newValue.locale('en').format('x'), 10);
  }

  return datePickerValue;
};

const addOffset = (useTimezone, endOfDay, value) => {
  const newValue = moment.utc(value);

  if (!useTimezone) {
    newValue.add(moment().utcOffset(), 'minute');
  }

  if (endOfDay) {
    const method = useTimezone ? newValue.local() : newValue.utc();
    method.endOf('day');
  }

  return newValue;
};

class DatePicker extends Component {
  constructor(props) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(datePickerValue) {
    const { endOfDay, useTimezone, onChange } = this.props;

    if (!datePickerValue) {
      onChange(null);
    } else {
      const newValue = addOffset(useTimezone, endOfDay, datePickerValue);
      onChange(parseInt(newValue.locale('en').format('X'), 10));
    }
  }

  render() {
    let { locale, format } = this.props;
    const { useTimezone, value } = this.props;
    locale = locale || 'en';
    format = format || 'dd/MM/yyyy';

    const datePickerValue = removeOffset(useTimezone, value);

    return (
      <DatePickerComponent
        dateFormat={format}
        className="form-control"
        onChange={this.handleChange}
        selected={datePickerValue}
        locale={locale}
        placeholderText={format}
        isClearable
        fixedHeight
        showYearDropdown
      />
    );
  }
}

DatePicker.defaultProps = {
  value: undefined,
  endOfDay: false,
  locale: undefined,
  format: undefined,
  useTimezone: false,
};

DatePicker.propTypes = {
  onChange: PropTypes.func.isRequired,
  value: PropTypes.number,
  endOfDay: PropTypes.bool,
  locale: PropTypes.string,
  format: PropTypes.string,
  useTimezone: PropTypes.bool,
};

export default connect()(DatePicker);
