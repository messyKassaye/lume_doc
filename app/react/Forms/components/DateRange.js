import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { t } from 'app/I18N';
import DatePicker from './DatePicker';

class DateRange extends Component {
  onChange(prop, propValue) {
    const { value, onChange } = this.props;
    const state = { ...value, [prop]: propValue };

    onChange(state);
  }

  render() {
    const { useTimezone, locale, format } = this.props;
    const { value } = this.props;
    const { from: stateFrom, to: stateTo } = value || { from: null, to: null };

    return (
      <div>
        <div className="DatePicker__From">
          <span>{t('System', 'Label date "From"', 'From:')}&nbsp;</span>
          <DatePicker
            locale={locale}
            format={format}
            useTimezone={useTimezone}
            value={stateFrom}
            onChange={val => this.onChange('from', val)}
          />
        </div>
        <div className="DatePicker__To">
          <span>&nbsp;{t('System', 'Label date "to"', 'To:')}&nbsp;</span>
          <DatePicker
            locale={locale}
            format={format}
            useTimezone={useTimezone}
            value={stateTo}
            endOfDay
            onChange={val => this.onChange('to', val)}
          />
        </div>
      </div>
    );
  }
}

DateRange.defaultProps = {
  value: { from: null, to: null },
  onChange: () => {},
  locale: undefined,
  format: undefined,
  useTimezone: false,
};

DateRange.propTypes = {
  value: PropTypes.shape({ from: PropTypes.number, to: PropTypes.number }),
  onChange: PropTypes.func,
  locale: PropTypes.string,
  format: PropTypes.string,
  useTimezone: PropTypes.bool,
};

export default DateRange;
