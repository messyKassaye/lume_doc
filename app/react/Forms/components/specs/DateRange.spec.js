/** @format */

import React from 'react';
import { shallow } from 'enzyme';

import DateRange from '../DateRange';
import DatePicker from '../DatePicker';

describe('DateRange', () => {
  let component;
  let props;

  const render = () => {
    component = shallow(<DateRange {...props} />);
  };

  beforeEach(() => {
    props = {
      onChange: jasmine.createSpy('onChange'),
      value: { from: 0, to: 1 },
    };

    render();
  });

  it('should render the component', () => {
    expect(component).toMatchSnapshot();
  });

  it('should allow using the local timezone', () => {
    props.useTimezone = true;
    render();
    expect(component).toMatchSnapshot();
  });

  describe('when a date is selected', () => {
    it('should triger onChange events', () => {
      component
        .find(DatePicker)
        .first()
        .simulate('change', 1469656800);
      expect(props.onChange).toHaveBeenCalledWith({ from: 1469656800, to: 1 });
      // We're mocking out onChange, so the change is not persisted.
      component
        .find(DatePicker)
        .last()
        .simulate('change', 1469656800);
      expect(props.onChange).toHaveBeenCalledWith({ from: 0, to: 1469656800 });
    });
  });
});
