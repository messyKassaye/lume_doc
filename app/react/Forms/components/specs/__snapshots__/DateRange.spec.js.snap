// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateRange should allow using the local timezone 1`] = `
<div>
  <div
    className="DatePicker__From"
  >
    <span>
      From:
       
    </span>
    <Connect(DatePicker)
      onChange={[Function]}
      useTimezone={true}
      value={0}
    />
  </div>
  <div
    className="DatePicker__To"
  >
    <span>
       
      To:
       
    </span>
    <Connect(DatePicker)
      endOfDay={true}
      onChange={[Function]}
      useTimezone={true}
      value={1}
    />
  </div>
</div>
`;

exports[`DateRange should render the component 1`] = `
<div>
  <div
    className="DatePicker__From"
  >
    <span>
      From:
       
    </span>
    <Connect(DatePicker)
      onChange={[Function]}
      useTimezone={false}
      value={0}
    />
  </div>
  <div
    className="DatePicker__To"
  >
    <span>
       
      To:
       
    </span>
    <Connect(DatePicker)
      endOfDay={true}
      onChange={[Function]}
      useTimezone={false}
      value={1}
    />
  </div>
</div>
`;
